input: add a user_data pointer to the callback, add a custom name callback

fabiobaltieri opened the pull request:
Add a user_data pointer to the callback structure and add a NAMED variant of the macro for multiple instance drivers.

This needs an update in the migration file as well, which is not there yet, will do it separately, I have something else to add there already anyway.

fabiobaltieri committed:
input: add a user_data pointer to the callback

Add a void *user_data pointer to the input callback structure. This is
useful for driver to get back the driver data structure and avoid
defining wrapper functions.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri committed:
input: add a INPUT_CALLBACK_DEFINE_NAMED macro

Add a INPUT_CALLBACK_DEFINE_NAMED macro, same as INPUT_CALLBACK_DEFINE
but with custom struct name, useful for declaring multiple callbacks
with different user_data on the same callback function.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri committed:
input: longpress, keymap: use an explicit callback name

Use an explicit callback name so that multiple instances of this do not
result in a:

redefinition of '_input_callback__longpress_cb'

error. This used to work when it was using unique generated wrappers,
but now it needs an index in the callback name.

Use it in one of the API tests as well, just in case.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

faxe1008 commented on samples/subsys/usb/hid-keyboard/src/main.c:
```diff
@@ -55,7 +55,7 @@ static uint8_t __aligned(sizeof(void *)) report[KB_REPORT_COUNT];
-static void input_cb(struct input_event *evt)
+static void input_cb(struct input_event *evt, void *user_data)
 {
 	struct kb_event kb_evt;
```
`ARG_UNUSED` missing here?

fabiobaltieri replied:
added

faxe1008 commented on samples/subsys/usb/hid-mouse/src/main.c:
```diff
@@ -55,7 +55,7 @@ static ALWAYS_INLINE void rwup_if_suspended(void)
-static void input_cb(struct input_event *evt)
+static void input_cb(struct input_event *evt, void *user_data)
 {
 	static uint8_t tmp[MOUSE_REPORT_COUNT];
```
`ARG_UNUSED` also.

fabiobaltieri replied:
added

faxe1008 commented on subsys/input/input_utils.c:
```diff
@@ -51,7 +51,7 @@ static bool input_dump_enabled(void)
-static void input_dump_cb(struct input_event *evt)
+static void input_dump_cb(struct input_event *evt, void *user_data)
 {
 	if (!input_dump_enabled()) {
```
`ARG_UNUSED`

fabiobaltieri replied:
added

faxe1008 commented on subsys/input/input_utils.c:
```diff
@@ -162,7 +162,7 @@ static void kbd_matrix_state_log_entry(char *header, kbd_row_t *data)
 		kbd_matrix_state_dev->name, header, kbd_matrix_buf, count);
 }
 
-static void kbd_matrix_state_log(struct input_event *evt)
```
`ARG_UNUSED`

fabiobaltieri replied:
added

faxe1008 reviewed (dismissed):
Wondering why there is no unused var complaint by the CI. Otherwise looks good to me  :^). I guess if you wanted to you could also add a test for the `NAMED` macro, but not needed.

bjarki-andreasen reviewed (commented):
Looks good :) I only noticed the missing `ARG_UNUSED()` in some functions

fabiobaltieri commented:
Added a bunch of `ARG_UNUSED`. Do you find them useful? Honestly I never bothered.

> Wondering why there is no unused var complaint by the CI

I don't think the check is enabled in CI, well I don't even know how to enable it locally, but yeah is this even part of the coding guideline? Anyway, you folks cared enough to point it out so I've added them.

fabiobaltieri commented:
> I guess if you wanted to you could also add a test for the `NAMED` macro, but not needed.

Yeah it's used by few drivers which are built in test_all, no harm adding it to one of the API test cases though, done.

tmon-nordic commented:
What does "custion" mean in "but with custion struct name"?

fabiobaltieri committed:
doc: migration-guide: add few input points

Add a point for the updated callback API as well as the analog axis
inversion property rename.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented on include/zephyr/input/input.h:
```diff
@@ -124,9 +124,26 @@ struct input_callback {
 	const struct device *dev;
 	/** The callback function. */
-	void (*callback)(struct input_event *evt);
+	void (*callback)(struct input_event *evt, void *user_data);
```
yeah was missing the template, merged and added

fabiobaltieri commented on include/zephyr/input/input.h:
```diff
@@ -124,9 +124,26 @@ struct input_callback {
 };
 
+/**
+ * @brief Register a callback structure for input events with an custom name.
```
thanks, fixed

fabiobaltieri commented:
> What does "custion" mean in "but with custion struct name"?

@tmon-nordic it means I can't type.

fabiobaltieri commented:
Rebased, added a migration guide and fixed few typos

kartben commented on doc/releases/migration-guide-4.0.rst:
```diff
@@ -54,6 +54,14 @@ GNSS
+  argument has to be added to the callback function arguments.
+
+* The :dtcompatible:`analog-axis` ``invert`` property has been renamed to
+  ``invert-input`` (there's now a ``invert-output`` available as well.
```
```suggestion
  ``invert-input`` (there's now an ``invert-output`` available as well).
```

fabiobaltieri replied:
fixed

kartben replied:
but you forgot the closing parenthesis boohooh

fabiobaltieri replied:
Ahhh!!!! AI HELP ME!
